packages/dd-trace/src/appsec/rasp/sql_injection.js

```diff
@@ -0,0 +1,90 @@
+'use strict'
+
+const { pgQueryStart, pgPoolQueryStart, wafRunFinished } = require('../channels')
+const { storage } = require('../../../../datadog-core')
+const addresses = require('../addresses')
+const waf = require('../waf')
+const { RULE_TYPES, handleResult } = require('./utils')
+
+const DB_SYSTEM_POSTGRES = 'postgresql'
+const reqQueryMap = new WeakMap() // WeakMap<Request, Set<querystring>>
+
+let config
+
+function enable (_config) {
+  config = _config
+
+  pgQueryStart.subscribe(analyzePgSqlInjection)
+  pgPoolQueryStart.subscribe(analyzePgSqlInjection)
+  wafRunFinished.subscribe(clearQuerySet)
+}
+
+function disable () {
+  if (pgQueryStart.hasSubscribers) pgQueryStart.unsubscribe(analyzePgSqlInjection)
+  if (pgPoolQueryStart.hasSubscribers) pgPoolQueryStart.unsubscribe(analyzePgSqlInjection)
+  if (wafRunFinished.hasSubscribers) wafRunFinished.unsubscribe(clearQuerySet)
+}
+
+function analyzePgSqlInjection (ctx) {
+  const query = ctx.query?.text
+  if (!query) return
+
+  const store = storage.getStore()
+  if (!store) return
+
+  const { raspSqlAnalyzed, req, res } = store
+
+  if (!req || raspSqlAnalyzed) return
+
+  let executedQueries = reqQueryMap.get(req)
+  if (executedQueries?.has(query)) return
+
+  // Do not waste time executing same query twice
+  // This also will prevent double calls in pg.Pool internal queries
+  if (!executedQueries) {
+    executedQueries = new Set()
+    reqQueryMap.set(req, executedQueries)
+  }
+  executedQueries.add(query)
+
+  const persistent = {
+    [addresses.DB_STATEMENT]: query,
+    [addresses.DB_SYSTEM]: DB_SYSTEM_POSTGRES
+  }
+
+  const result = waf.run({ persistent }, req, RULE_TYPES.SQL_INJECTION)
+
+  handleResult(result, req, res, ctx.abortController, config)
+}
+
+function hasInputAddress (payload) {
+  let appliedAddresses = []
+  if (payload.ephemeral) {
+    appliedAddresses = appliedAddresses.concat(Object.keys(payload.ephemeral))
+  }
+  if (payload.persistent) {
+    appliedAddresses = appliedAddresses.concat(Object.keys(payload.persistent))
+  }
```

Comment:
Less optimized than iterating but more compact and fancy :P
```suggestion
function hasInputAddress ({ ephemeral = [], persistent= [] }) {
  const appliedAddresses = [...Object.keys(ephemeral), ...Object.keys(persistent)]
```

Reply:
This was already modified when you send the message, now the code is different.